pai-management/k8sPaiLibrary/maintainlib/common.py

```diff
@@ -167,11 +167,16 @@ def sftp_paramiko(src, dst, filename, host_config):
         if port_validation(host_config['sshport']) == False:
             return False
         port = int(host_config['sshport'])
-
+    key_filename = None
+    if 'key-filename' in host_config:
+        if os.path.isfile(str(host_config['key-filename'])) and host_config['key-filename'] is not None:
+            key_filename = str(host_config['key-filename'])
+        else:
+            logger.warn("The key file: {0} specified doesn't exist".format(host_config['key-filename']))
     # First make sure the folder exist.
     ssh = paramiko.SSHClient()
     ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
-    ssh.connect(hostname=hostip, port=port, username=username, password=password)
+    ssh.connect(hostname=hostip, port=port, key_filename=key_filename, username=username, password=password)
```

Comment:
>key_filename=key_filename, username=username, password=password [](start = 44, length = 63)

If both key and password exists, which one has higher priority?

Reply:
```
Authentication is attempted in the following order of priority:

The pkey or key_filename passed in (if any)
key_filename may contain OpenSSH public certificate paths as well as regular private-key paths; when files ending in -cert.pub are found, they are assumed to match a private key, and both components will be loaded. (The private key itself does not need to be listed in key_filename for this to occur - just the certificate.)
Any key we can find through an SSH agent
Any “id_rsa”, “id_dsa” or “id_ecdsa” key discoverable in ~/.ssh/
When OpenSSH-style public certificates exist that match an existing such private key (so e.g. one has id_rsa and id_rsa-cert.pub) the certificate will be loaded alongside the private key and used for authentication.
Plain username/password auth, if a password was given
```

REF: http://docs.paramiko.org/en/2.4/api/client.html